File: src/data/4o/Largest_Rectangle_in_Histogram/Largest_Rectangle_in_Histogram.java
import java.util.Stack;

class Solution {
    public int largestRectangleArea(int[] heights) {
        int maxArea = 0;
        Stack<Integer> stack = new Stack<>();
        int n = heights.length;

        for (int i = 0; i <= n; i++) {
            int h = (i == n) ? 0 : heights[i];
            while (!stack.isEmpty() && h < heights[stack.peek()]) {
                int height = heights[stack.pop()];
                int width = stack.isEmpty() ? i : i - stack.peek() - 1;
                maxArea = Math.max(maxArea, height * width);
            }
            stack.push(i);
        }

        return maxArea;
    }
}

public class Largest_Rectangle_in_Histogram {
public static void runTests(Solution solution) {
    // Test case 1: Simple case with increasing heights
    int[] heights1 = {2, 1, 5, 6, 2, 3};
    int expected_result1 = 10;
    assert solution.largestRectangleArea(heights1) == expected_result1 : "Test case 1 failed. Expected " + expected_result1;

    // Test case 2: All bars of the same height
    int[] heights2 = {2, 2, 2, 2, 2};
    int expected_result2 = 10;
    assert solution.largestRectangleArea(heights2) == expected_result2 : "Test case 2 failed. Expected " + expected_result2;

    // Test case 3: Single bar
    int[] heights3 = {5};
    int expected_result3 = 5;
    assert solution.largestRectangleArea(heights3) == expected_result3 : "Test case 3 failed. Expected " + expected_result3;

    // Test case 4: Decreasing bar heights
    int[] heights4 = {6, 5, 4, 3, 2, 1};
    int expected_result4 = 12;
    assert solution.largestRectangleArea(heights4) == expected_result4 : "Test case 4 failed. Expected " + expected_result4;

    // Test case 5: Bars with alternating heights
    int[] heights5 = {1, 3, 2, 1, 2};
    int expected_result5 = 5;
    assert solution.largestRectangleArea(heights5) == expected_result5 : "Test case 5 failed. Expected " + expected_result5;

    // Test case 6: Empty histogram (no bars)
    int[] heights6 = {};
    int expected_result6 = 0;
    assert solution.largestRectangleArea(heights6) == expected_result6 : "Test case 6 failed. Expected " + expected_result6;

    // Test case 7: Bars with large values
    int[] heights7 = {1000, 1000, 1000, 1000};
    int expected_result7 = 4000;
    assert solution.largestRectangleArea(heights7) == expected_result7 : "Test case 7 failed. Expected " + expected_result7;
}

    public static void main(String[] args) {
        Solution solution = new Solution();
        for (int i = 0; i < 100000000; i++) {
            runTests(solution); // Ensure your test cases are implemented inside runTests
        }
    }
}

    

File: src/data/4o/Largest_Rectangle_in_Histogram/Largest_Rectangle_in_Histogram.cpp
#include <vector>
#include <string>
using namespace std;

class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        int n = heights.size();
        stack<int> st;
        int maxArea = 0;

        for (int i = 0; i <= n; i++) {
            int h = (i == n) ? 0 : heights[i];
            while (!st.empty() && h < heights[st.top()]) {
                int height = heights[st.top()];
                st.pop();
                int width = st.empty() ? i : i - st.top() - 1;
                maxArea = max(maxArea, height * width);
            }
            st.push(i);
        }

        return maxArea;
    }
};

//  Test cases for the C++ solution
#include <cassert>
// Test cases for the C++ solution
void testSolution(Solution& solution) {
    // Test case 1: Simple case with increasing heights
    std::vector<int> heights = {2, 1, 5, 6, 2, 3};
    int expected_result = 10;
    assert(solution.largestRectangleArea(heights) == expected_result && "Test case 1 failed.");

    // Test case 2: All bars of the same height
    heights = {2, 2, 2, 2, 2};
    expected_result = 10;
    assert(solution.largestRectangleArea(heights) == expected_result && "Test case 2 failed.");

    // Test case 3: Single bar
    heights = {5};
    expected_result = 5;
    assert(solution.largestRectangleArea(heights) == expected_result && "Test case 3 failed.");

    // Test case 4: Decreasing bar heights
    heights = {6, 5, 4, 3, 2, 1};
    expected_result = 12;
    assert(solution.largestRectangleArea(heights) == expected_result && "Test case 4 failed.");

    // Test case 5: Bars with alternating heights
    heights = {1, 3, 2, 1, 2};
    expected_result = 5;
    assert(solution.largestRectangleArea(heights) == expected_result && "Test case 5 failed.");

    // Test case 6: Empty histogram (no bars)
    heights = {};
    expected_result = 0;
    assert(solution.largestRectangleArea(heights) == expected_result && "Test case 6 failed.");

    // Test case 7: Bars with large values
    heights = {1000, 1000, 1000, 1000};
    expected_result = 4000;
    assert(solution.largestRectangleArea(heights) == expected_result && "Test case 7 failed.");
}


int main() {
    Solution solution;
    for (int i = 0; i < 1310720; ++i) {
        testSolution(solution);  
    }
    return 0;
}
